### ISSUE TITLE
scipy.signal needs a unified filter API

### ISSUE THREAD

**Issue by e-q** [scipy/scipy#6137]
Here I will present some thoughts I have collected about the status of filtering and LTI system analysis in scipy.signal. This issue will hopefully lead to a discussion that can address factors that have frequently arisen in multiple PRs, but are generally of broader scope than can be settled in a single PR. While I have almost certainly overlooked relevant considerations, much of what follows was informed by reviewing the following discussions:

gh-3259, gh-3717, gh-4576, gh-4881

We should keep the following use cases in mind:

**Control systems**: Design a continuous-time LTI controller, analyze stability, simulate continuous time responses
**DSP**: Design discrete time FIR/IIR filters, simulate discrete time responses, apply filters, resample
**Non-expert use**: Just want to lowpass some data in a simple way

I welcome any and all feedback and opinions!

# Problems:
- The number of filter/system representations has led to some disarray:
    - Current call signatures use a mishmash of system-tuples, lti objects, and individual coefficient arguments
    - Conversions between representations are not always optimal for a given use case
    - We mostly default to `(b, a)` form, which is numerically unstable for anything beyond low filter orders
- Inconsistent specification of sampling time/freq in many functions
- Inconsistent specification of angular or regular frequency units in many functions. 

# Possibilities
## Object focused
We have added continuous time LTI objects, and a PR for discrete time objects is well underway. One option is to standardize all of the filter/LTI parts of the module on these LTI objects. I think that the main drawback of this approach would be unfamiliarity to MATLAB/Octave converts, who expect the `lfilter(b, a, data)` style syntax. 

There are a few ways this could go.

One vision, which @endolith has advocated, is a system where the filter representation is mostly hidden from the user. Behind the scenes, the LTI object's methods would automatically use the right representation for the job, and convert as necessary. This allows things like

```python
    fs = 48000
    signal = randn(fs)
    fc = 2000
    lowpass = butter(4, fc, fs=fs, analog=False, output='lti')
    w, h = lowpass.freqresp()  # Uses ZPK internally
    t, y = lowpass.dstep()  # Uses SS internally
    num = lowpass.num  # Automatically calculates TF form internally
```

Something similar, as @befelix has suggested, would not include automatic conversion, the only instantiable classes are LTI subclasses that define their representation. This would also allow the LTI subclasses to have methods unique to them. The methods would largely be the same, and methods could still internally convert. For example:

```python
    lowpass_zpk = butter(4, fc, fs=fs, analog=False, output='zpk')
    t, y = dstep(lowpass_zpk)  # Converts to SS inside of dstep
    zeros = lowpass_zpk.zeros
    num = lowpass_zpk.as_tf().num
```

In each of these scenarios, there are multiple possibilities for how a given filter may be applied. For example, there could be one unified `filter` function that allows different methods (which may be a method of the LTI object and/or accept an LTI object as an argument)

```python
    filter(system, data, method='lfilter')
    filter(system, data, method='filtfilt')
    filter(system, data, method='sosfilt')
```

Or, the filtering functions could take LTI objects as their arguments:

```python
    lfilter(system, data)
    filtfilt(system, data)
    sosfilt(system, data)
```

## Coefficient focused
Alternatively, maybe people don't want to move towards an object-heavy model. In this paradigm, call signatures would be limited to arguments for specific kinds of coefficients, or use the system-tuple approach.

```python
    z, p, k = butter(4, fc, fs=fs, analog=False, output='zpk')
    b, a = zpk2tf(z, p, k)
    w, mag, ph = bode((z, p, k))
    w, h = freqz(b, a)
    t, out = dstep((A,B,C,D,dt))
    filtered = lfilter(b, a, x)
```

This is more involved than the object-focused approach, but more familiar to users coming from other languages. 
    
# Consistency
Regardless of what API style is agreed up, consistency is of utmost importance. Once an agreement it achieved, all of the filter design functions should be standardized in terms of their arguments.

For example:
- Are discrete-time system frequencies normalized by the nyquist or sampling frequency?
- Should the user specify sampling frequency via `fs` or sample time via `dt`? It has been suggested that `fs` is more likely to be an exact integer, so we can avoid testing equality of floats when comparing rates.
- Do functions like `bode`, `freqresp` return a frequency array in  angular or normal frequency units?

# Open Questions

- What API seems like the best option?
- It'll be a long slog to get to whatever vision is agreed upon with the usual deprecation procedure. What prospect exists for making a clean break?
- What pros/cons do people see in the suggestions above? 
- What other forms could the API take?

**Comment by endolith** [comment]
Actually, now that `TransferFunction` and `ZerosPolesGain` exists as distinct objects, in https://github.com/scipy/scipy/issues/5613#issuecomment-215930824 I agreed that it makes more sense that calling a TransferFunction should give the H(s) response.  And since the "callable filter" object would use ZPK in some places and SOS in other places, it should probably be a distinct DigitalFilter object, so like this:

```
sig = randn(48000)
# 2 kHz lowpass DigitalFilter object
lowpass = butter(4, 2000, fs=48000, analog=False, output='object')
filtered_sig = lowpass(sig)
```

whether it should also have `.freqresp`, `.dstep`, etc., I haven't thought about it.

**Comment by nbecker** [comment]
I can speak of my personal experience, as DSP is a major part of my work.  I have developed my own tools over many years, which use boost:python + c++. FIR filters are objects, because Anything that needs to have internal state is an object.

I have separate classes for plain FIR filter, decimating filter, and interpolating filter.  This is necessary to implement interpolation and decimation efficiently.  Filters are also segregated according to data types of input and coefficient (float vs. complex).

All filters are designed to maintain state so that inputs can be fed in chunks.  In particular, decimating filters need to maintain an internal phase.

I also have code for efficient rational rate conversion, but this is not directly related code.

The only drawback to this approach that I see is it's unfamiliar to those coming from Matlab.  In my opinion that approach is inadequate for my needs and supporting it would be a low priority.

**Comment by Eric89GXL** [comment]
> since the "callable filter" object would use ZPK in some places and SOS in other places, it should probably be a distinct DigitalFilter object

This makes sense to me. @e-q do you want to update the issue comment at the top to reflect this change?

> I also have code for efficient rational rate conversion, but this is not directly related code.

FYI as of a couple of months ago we have this now with `resample_poly`, which uses a Cython implementation of `upfirdn`. The `upfirdn` uses internally a class that could be expanded to do the state tracking you mention. PR welcome if you're up for it.

I suspect we could support both simple interfaces (MATLAB-like) and advanced interfaces (chunk-feeding object) for e.g. filtering.

Once the advanced objects exist, we can use them in the simpler MATLAB-like interfaces without too much extra overhead. We already have the MATLAB-like interfaces and they are likely widely used. I don't think we should get rid of them, given their low maintenance overhead (once the underlying advanced objects are available).

This is more or less how I envisioned the `resample_poly`->`upfirdn`->`_UpFIRDn` hierarchy shaping up going forward. @SirTom67's SWIG code actually originally had internal states built into it, but I took them out to get things working quickly for `scipy`. My hope is that some contributors might come along to make e.g. `_UpFIRDn` do state tracking with proper tests, scipy-like implementation, etc. so we can make it public.

In any case, I do like @endolith's idea of making callable classes that take care of "the right way" of doing things if possible. This will interact with the idea of being able to feed in data in chunks, though. Someone might rightly expect `x_lp = lowpass(x)` followed by `y_lp = lowpass(y)` to work just fine, but if we do chunk-based processing by default these will be treated like two chunks from the same system. Maybe we could have `maintain_state=False` by default in the constructor, which would mean "treat all inputs as independent" (I think most people will use it this way) or `maintain_state=True` to allow for `x_lp = np.concatenate((lowpass(x[:10]), lowpass(x[10:]))` to give the same thing as `x_lp = lowpass(x)`.

**Comment by endolith** [comment]
> filters are designed to maintain state so that inputs can be fed in chunks.

Good point,  another reason to separate the filter objects from the transfer function objects.  So they'd also have a method/argument to set the initial state/conditions?

**Comment by FRidh** [comment]
I like to keep filter description and the actual filtering separated from each other. I agree that filters should be seen as objects because they have state and properties. I do would like to keep the actual filtering in separate functions, and have methods that refer to these functions. 

Preferably something like singledispatch is used in these functions so it is possible to implement custom filter functions while still using the filter objects. The use case I have in mind is a module I am working on that uses iterators instead of arrays (https://github.com/FRidh/streaming). That would also cover I think the following case.

> In any case, I do like @endolith's idea of making callable classes that take care of "the right way" of doing things if possible. This will interact with the idea of being able to feed in data in chunks, though. Someone might rightly expect x_lp = lowpass(x) followed by y_lp = lowpass(y) to work just fine, but if we do chunk-based processing by default these will be treated like two chunks from the same system. Maybe we could have maintain_state=False by default in the constructor, which would mean "treat all inputs as independent" (I think most people will use it this way) or maintain_state=True to allow for x_lp = np.concatenate((lowpass(x[:10]), lowpass(x[10:])) to give the same thing as x_lp = lowpass(x).

Furthermore, I think it makes sense to create either a new top-level module or else a separate project/scikit as I believe it allows faster development and we wouldn't have to worry about backwards compatibility as much.

```py

# Abstract base classes

class System(object):
    pass

class LinearTimeInvariant(System):

class Digital(System):

    @property
    def isdigital(self):
        return True

    def fs(self):
        # Sampling frequency.
        # We should emphasize that frequency is 1/x where x can be any dimension, not just time.
        return self._fs

class Analog(System):

    @property
    def isdigital(self):
        return False

class Representation(LinearTimeInvariant):

    def __iter__(self):
        pass

    @abstractproperty
    def representation(self):
        pass

    def to(self, representation):
        """Uses :func:`convert`"""
        pass


    def filter(self, x, axis=-1, zi=None):
        # uses self.representation and :func:`filter`
        pass

class TransferFunction(Representation):

    @property
    def representation(self):
        return 'tf'

class SecondOrderSections(Representation):
    pass

class ZerosPolesGain(Representation):
    pass

class StateSpace(Representation):
    pass

# Instantiable classes

class TransferFunctionAnalog(TransferFunction, Analog):
    pass

class TransferFunctionDigital(TransferFunction, Digital):
    pass

class SecondOrderSectionsAnalog(LinearTimeInvariant, Analog):
    pass

class SecondOrderSectionsDigital(LinearTimeInvariant, Digital):
    pass

class ZerosPolesGainAnalog(LinearTimeInvariant, Analog):
    pass

class ZerosPolesGainDigital(LinearTimeInvariant, Digital):
    pass

class StateSpaceAnalog(LinearTimeInvariant, Analog):
    pass

class StateSpaceDigital(LinearTimeInvariant, Digital):
    pass

REPRESENTATIONS = ['tf', 'zpk', 'sos', 'ss']

FILTERS = {
    'analog'    : {
        'tf'    :   TransferFunctionAnalog,
    },
    'digital'   : {
        'tf'    :   TransferFunctionDigital,
        }
    }

FILTER_FUNCTIONS = {
    'tf'    :   filter_ba,
    }

# In the following functions `system` is a tuple, not a Filter object
# We can use singledispatch to allow filters that work not on just sequences, but also on e.g. iterators.

@singledispatch
def filter(x, system, representation, axis=-1, zi=None):
    # This function only returns the right function.
    # It should not touch `x`.
    pass

@singledispatch
def filter_ba(x, ba, axis=-1, zi=None):
    pass

@singledispatch
def filter_sos(x, sos, axis=-1, zi=None):
    pass

@singledispatch
def filter_zpk(x, zpk, axis=-1, zi=None):
    pass

@singledispatch
def filter_ss(x, ss, axis=-1, zi=None):
    pass

@singledispatch
def filtfilt(x, system, representation, ...,):
    # Filter both directions
    pass

@singledispatch
def filtfilt_ba(x, ba, axis=-1, ...):
    pass

def convert(system, current, new):
    """Convert representation. This function calls one of the functions below."""
    pass

def tp2zpk():
    pass

def tf2sos():
    pass

def tf2ss():
    pass

...
```

**Comment by nbecker** [comment]
I'm not a fan of the overuse of inheritance.  In my own work, I find that it rarely serves any purpose in python.  Unless those base classes are really adding value, I'd suggest a flatter design.

**Comment by FRidh** [comment]
@nbecker It depends on how we would store the coefficients and make them accessible. It is indeed possible to get rid of the classes for each system type, if we choose to store coefficients as a tuple, and not have properties referring to the items in the tuple.

In my example `System` and `LinearTimeInvariant` could be merged because we do not support time-variant systems. Maybe we want to, but that makes the problem significantly more difficult.

```py
# Abstract base class
class System(object):

    def __iter__(self):
        yield from self.system

    @abstractproperty
    def representation(self):
        return self._representation # e.g. 'tf'

    @abstractproperty
    def system(self):
        return self._system

    def to(self, representation):
        """Uses :func:`convert`"""
        pass


    def filter(self, x, axis=-1, zi=None):
        # uses self.representation and :func:`filter`
        pass

# Instantiable classes

class DigitalFilter(System):

    def __init__(self, system, fs):
        pass

    @property
    def isdigital(self):
        return True

    def fs(self):
        # Sampling frequency.
        # We should emphasize that frequency is 1/x where x can be any dimension, not just time.
        return self._fs

class AnalogFilter(System):

    def __init__(self, system):
        pass

    @property
    def isdigital(self):
        return False
```

**Comment by FRidh** [comment]
It would also be useful to have a function `cascade` that takes multiple `Filter` objects and returns a `Filter` that represents the cascaded filter.

**Comment by rgommers** [comment]
> It'll be a long slog to get to whatever vision is agreed upon with the usual deprecation procedure. What prospect exists for making a clean break?

The current API is well established and very heavily used, so if there are significant changes then a clean break is much preferred I'd think.

Depending on what API rolls out of this: a separate submodule within ``signal``, or if it's a limited number of classes with distinct enough names, keep those in the ``signal`` namespace.

> new top-level module or else a separate project/scikit as I believe it allows faster development and we wouldn't have to worry about backwards compatibility as much.

A new top-level module is quite unlikely to happen. Something like ``signal.filtering`` can be done. A separate package is also an option of course (or develop standalone, and then with some git surgery bolt it back onto ``signal``).

**Comment by drestebon** [comment]
I want to point out that in scipy 1.11.2 the conventions in `TransferFunction` and the filter design functions are not compatible.

There is at least two conventions. In both of them the coefficients are put in an array ordered with descending powers of s, or z. In `TransferFunction` the first element is the coefficient of the highest power (what that is depends on the lengths of the arrays). In lfilter the first element is the coefficient of the power 0.

This leads to confusing behavior (although expected after reading the documentation).

For illustration see below. `dimpulse` and passing a unit impulse through `lfilter` yield different results, for the same sets of coefficients:

```
import numpy as np
import scipy.signal as sig
n = 6
a = sig.unit_impulse(n)
for l,f in [
        ('lfilter', lambda x : sig.lfilter(x,a,a)),
        ('dimpulse', lambda x : sig.dimpulse((x,a,1),n=n)[1][0].flatten()) ]:
    print(l)
    for blen in range(1,n+1):
        b = sig.unit_impulse(blen)
        i = f(b)
        print('ir = ', i, 'a =', a, 'b =',b)
```

the output:

```
lfilter
ir =  [1. 0. 0. 0. 0. 0.] a = [1. 0. 0. 0. 0. 0.] b = [1.]
ir =  [1. 0. 0. 0. 0. 0.] a = [1. 0. 0. 0. 0. 0.] b = [1. 0.]
ir =  [1. 0. 0. 0. 0. 0.] a = [1. 0. 0. 0. 0. 0.] b = [1. 0. 0.]
ir =  [1. 0. 0. 0. 0. 0.] a = [1. 0. 0. 0. 0. 0.] b = [1. 0. 0. 0.]
ir =  [1. 0. 0. 0. 0. 0.] a = [1. 0. 0. 0. 0. 0.] b = [1. 0. 0. 0. 0.]
ir =  [1. 0. 0. 0. 0. 0.] a = [1. 0. 0. 0. 0. 0.] b = [1. 0. 0. 0. 0. 0.]
dimpulse
ir =  [0. 0. 0. 0. 0. 1.] a = [1. 0. 0. 0. 0. 0.] b = [1.]
ir =  [0. 0. 0. 0. 1. 0.] a = [1. 0. 0. 0. 0. 0.] b = [1. 0.]
ir =  [0. 0. 0. 1. 0. 0.] a = [1. 0. 0. 0. 0. 0.] b = [1. 0. 0.]
ir =  [0. 0. 1. 0. 0. 0.] a = [1. 0. 0. 0. 0. 0.] b = [1. 0. 0. 0.]
ir =  [0. 1. 0. 0. 0. 0.] a = [1. 0. 0. 0. 0. 0.] b = [1. 0. 0. 0. 0.]
ir =  [1. 0. 0. 0. 0. 0.] a = [1. 0. 0. 0. 0. 0.] b = [1. 0. 0. 0. 0. 0.]
```

**Comment by sirtom67** [comment]
You're right, that is confusing!
I'm not sure what to do about it. And, I'm not sure this thread is still
relevant. It was opened 7 years ago, after all! That must be some kind of
record. :-)

**Comment by drestebon** [comment]
Yes. I wouldn't know what to do about it either. Maybe the mailing list would be a better place for this discussion?

The trade off is that the conventions for continuous and discrete time are actually (usually?) different. So the decision is between having each family of functions be internally consistent (current situation), or having domains (continuous or discrete) be consistent across different families of functions. I tend to think that the later is better for interoperability: you get the same result from `signal.lfilter` and `signal.dlsim` for the same set of coefficients.

Continuous time transfer functions are usually written using positive powers of `s`:

```
                     M          M-1            
       B(s)    b[0] s    + b[1]s    + ... + b[M]
G(s) = ---- = -----------------------------------
                     N          N-1            
       A(s)    a[0] s    + a[1]s    + ... + a[N]
```

This matches the convention used in `TransferFunction` (for both continuous and discrete systems).

On the other hand, discrete time transfer functions are usually written using negative powers of `z`:

```
                           -1              -M
       B(z)    b[0] + b[1]z   + ... + b[M]z
H(z) = ---- = --------------------------------
                           -1              -N
       A(z)    a[0] + a[1]z   + ... + a[N]z
```

This matches the contention used in `freqz`, `lfilter`, etc.
